src/mz/src/bin/mz/upgrader.rs

```diff
@@ -0,0 +1,213 @@
+        };
+
+        let serialized_data = serde_json::to_string(&cache_data)?;
+        let mut file = File::create(&cache_path)?;
+        file.write_all(serialized_data.as_bytes())?;
+
+        Ok(())
+    }
+
+    /// Checks if the last timestamp in the temp file has more than one week old.
+    fn is_cache_older_than_a_week(&self, stored_timestamp: SystemTime) -> Result<bool, Error> {
+        if SystemTime::now()
+            .duration_since(stored_timestamp)
+            .map_err(|_| Error::TimestampConversionError)?
+            >= time::Duration::weeks(1)
+        {
+            return Ok(true);
+        }
+        Ok(false)
+    }
+
+    /// Returns the modified date and the version from the cache file.
+    /// If the file does not exist, the function returns `Ok(None)`.
+    fn get_cached_timpestamp_and_version(&self) -> Result<Option<(SystemTime, Version)>, Error> {
+        let mut cache_path: PathBuf = dirs::cache_dir().ok_or(Error::CacheDirNotFoundError)?;
+        cache_path.push(CACHE_FILE_NAME);
+
+        if !Path::new(&cache_path).exists() {
+            // Trigger the check if the file does not exist.
+            return Ok(None);
+        }
+
+        let file = OpenOptions::new().read(true).open(&cache_path)?;
+        let modified_time = file.metadata()?.modified()?;
+
+        // Read cache file
+        let mut reader = BufReader::new(file);
+        let mut file_content = String::new();
+        reader.read_line(&mut file_content)?;
+
+        let cache_file: CacheFile = serde_json::from_str(&file_content)?;
+
+        Ok(Some((modified_time, cache_file.last_read_version)))
+    }
+
+    /// Returns true if the installed version of `mz`
+    /// is older than the version sent by parameter.
+    fn is_installed_version_older_than(&self, version: Version) -> bool {
+        let installed_version = build_info!().semver_version();
+
+        version > installed_version
+    }
+
+    /// Fetches and returns the latest tag version from the Materialize repository
+    /// using the GitHub public API.
+    async fn get_latest_tag_version(&self) -> Result<Version, Error> {
+        // The policy must be set to None.
+        // Otherwise the client will handle the redirection.
+        let client = Client::builder().redirect(Policy::none()).build()?;
+
+        // Set an agent. Otherwise the response is not useful.
+        // let mut headers = HeaderMap::new();
+        // let user_agent = format!("mz/{}", VERSION.clone());
+        // headers.insert(
+        //     USER_AGENT,
+        //     HeaderValue::from_str(&user_agent).map_err(Error::HeaderParseError)?,
+        // );
+        let response = client
+            .get(BINARIES_LATEST_VERSION_URL)
+            // .headers(headers)
+            .send()
+            .await?;
+
+        // If the server returns a redirect, you can usually find the location header.
+        if let Some(latest_version) = response.headers().get("mz-latest.version") {
```

Comment:
Ah, it's not going to be in the header but in the body. You should be able to do just this now:

```
let version = reqwest::get(BINARIES_LATEST_VERSION_URL).await?.text().await?;
```